### PR TITLE
Fix assertion/possible deadlock in thread shutdown in the presence of ThreadStatics on collectible assemblies

### PR DESCRIPTION
**Pull request by davidwrighton**
Move all the locked accesses of the handle table and free list to being under a leaf lock

### CODE REVIEW

**Review by mangod9 — APPROVED** [review]

File: src/coreclr/vm/object.h [hunk 1]
```diff
@@ -2081,6 +2081,7 @@ class LoaderAllocatorObject : public Object
     friend class CoreLibBinder;
 
 public:
+    // All uses of this must be protected by m_crstLoaderAllocatorHandleTable or be safe lock-free reads
     PTRARRAYREF GetHandleTable()
     {
         LIMITED_METHOD_DAC_CONTRACT;
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/object.h:2087]: Can we update the contract to validate the comment above?

File: src/coreclr/vm/object.h [hunk 2]
```diff
@@ -2099,6 +2100,7 @@ class LoaderAllocatorObject : public Object
         return m_slotsUsed;
     }
 
+    // All uses of this must be protected by m_crstLoaderAllocatorHandleTable
     void SetSlotsUsed(INT32 newSlotsUsed)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/object.h:2104]: Same comment as GetHandleTable?

File: src/coreclr/vm/loaderallocator.cpp [hunk 3]
```diff
@@ -781,7 +781,7 @@ LOADERHANDLE LoaderAllocator::AllocateHandle(OBJECTREF value)
                         retVal = (UINT_PTR)((freeHandleIndex + 1) << 1);
                         break;
                     }
-
+
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/loaderallocator.cpp:784]:
```suggestion

```

**Review by AaronRobinsonMSFT — APPROVED** [review]
Thanks!

**Pull request merged**